server/api/apiController.js

```diff
@@ -117,138 +118,178 @@ module.exports = {
     req.logout();
-    serveStatus(res, 204);
+    return serveStatus(res, 204);
   },
 
+  // Must be admin
   postAddMember: function(req, res, next) {
     dbUtils.getUser(req.user, function(error, user) {
       if (error) {
         console.error(error);
-        serveStatus(res, 500);
-      } else {
+        return serveStatus(res, 500);
+      }
+      dbUtils.getOrganization({id: user.organization_id}, function(error, org) {
+        if (error) {
+          console.error(error);
+          return serveStatus(res, 500);
+        }
+        if (user.id !== org.admin_id) {
+          return serveStatus(res, 403);
+        }
         req.body.organization_id = user.organization_id;
         dbUtils.addUser(req.body, function(error, user) {
           if (!checkUserError(res, error)) {
             return serveStatus(res, 201);
           }
         });
-      }
+      });
     });
   },
 
+  // Must be admin and share organizations or be updating self
   postUpdateMember: function(req, res, next) {
-    dbUtils.getUser(req.user, function(error, user) {
+    if (req.user.id === req.body.id) {
+      dbUtils.updateUser(req.body, function(error, admin) {
+        if (!checkUserError(res, error)) {
+          return serveStatus(res, 204);
+        }
+      });
+      return;
+    }
+    dbUtils.getUser(req.user, function(error, admin) {
       if (error) {
         console.error(error);
-        serveStatus(res, 500);
-      } else {
-        // Forbid admins from editing users not in their organization
-        if (!user.organization_id && user.organization_id !== req.body.organization_id) {
-          serveStatus(res, 403);
-        } else {
+        return serveStatus(res, 500);
+      }
+      dbUtils.getUser({id: req.body.id}, function(error, user) {
+        if (error) {
+          console.error(error);
+          return serveStatus(res, 500);
+        }
+        if (admin.organization_id !== user.organization_id) {
+          return serveStatus(res, 403);
+        }
+        dbUtils.getOrganization({id: admin.organization_id}, function(error, org) {
+          if (error) {
+            console.error(error);
+            return serveStatus(res, 500);
+          }
+          if (admin.id !== org.admin_id) {
+            return serveStatus(res, 403);
+          }
           dbUtils.updateUser(req.body, function(error, user) {
             if (!checkUserError(res, error)) {
               return serveStatus(res, 204);
             }
           });
-        }
-      }
+        });
+      });
     });
   },
 
+  // Must share organizations
   postPing: function(req, res, next) {
     var params = req.body;
-    dbUtils.getUser(params, function(error, user) {
+    dbUtils.getUser(req.user, function(error, user) {
       if (error) {
         console.error(error);
-        serveStatus(res, 500);
+        return serveStatus(res, 500);
+      }
+      if (!req.user.organization_id || req.user.organization_id !== params.organization_id) {
```

Comment:
I think you need to fetch this data, it's not provided by default. Example:

`req.user = { id: 9, admin_only: false }`
`params = { id: '9', visitor: '', text: '' }`